File: examples/Algorithm/python/25_network_flow.py
"""
Network Flow
Maximum Flow and Bipartite Matching

Algorithms for computing maximum flow in graphs.
"""

from typing import List, Tuple, Dict
from collections import defaultdict, deque


# =============================================================================
# 1. Ford-Fulkerson (DFS-Based)
# =============================================================================

def ford_fulkerson(capacity: List[List[int]], source: int, sink: int) -> int:
    """
    Ford-Fulkerson Algorithm (DFS)
    Time Complexity: O(E * max_flow)
    """
    n = len(capacity)
    residual = [row[:] for row in capacity]

    def dfs(node: int, flow: int, visited: List[bool]) -> int:
        if node == sink:
            return flow

        visited[node] = True

        for next_node in range(n):
            if not visited[next_node] and residual[node][next_node] > 0:
                min_flow = min(flow, residual[node][next_node])
                result = dfs(next_node, min_flow, visited)

                if result > 0:
                    residual[node][next_node] -= result
                    residual[next_node][node] += result
                    return result

        return 0

    max_flow = 0
    while True:
        visited = [False] * n
        flow = dfs(source, float('inf'), visited)
        if flow == 0:
            break
        max_flow += flow

    return max_flow


# =============================================================================
# 2. Edmonds-Karp (BFS-Based)
# =============================================================================

def edmonds_karp(capacity: List[List[int]], source: int, sink: int) -> int:
    """
    Edmonds-Karp Algorithm (BFS)
    Time Complexity: O(V * E^2)
    """
    n = len(capacity)
    residual = [row[:] for row in capacity]

    def bfs() -> List[int]:
        """Find augmenting path via BFS"""
        parent = [-1] * n
        visited = [False] * n
        visited[source] = True
        queue = deque([source])

        while queue:
            node = queue.popleft()

            for next_node in range(n):
                if not visited[next_node] and residual[node][next_node] > 0:
                    visited[next_node] = True
                    parent[next_node] = node
                    queue.append(next_node)

                    if next_node == sink:
                        return parent

        return parent

    max_flow = 0

    while True:
        parent = bfs()

        if parent[sink] == -1:
            break

        # Find minimum capacity along the path
        path_flow = float('inf')
        node = sink
        while node != source:
            prev = parent[node]
            path_flow = min(path_flow, residual[prev][node])
            node = prev

        # Update residual graph
        node = sink
        while node != source:
            prev = parent[node]
            residual[prev][node] -= path_flow
            residual[node][prev] += path_flow
            node = prev

        max_flow += path_flow

    return max_flow


# =============================================================================
# 3. Dinic's Algorithm
# =============================================================================

class Dinic:
    """
    Dinic's Algorithm
    Time Complexity: O(V^2 * E)
    For bipartite graphs: O(E * sqrt(V))
    """

    def __init__(self, n: int):
        self.n = n
        self.graph = defaultdict(list)

    def add_edge(self, u: int, v: int, cap: int):
        """Add edge (u -> v, capacity cap)"""
        # (adjacent node, residual capacity, reverse edge index)
        self.graph[u].append([v, cap, len(self.graph[v])])
        self.graph[v].append([u, 0, len(self.graph[u]) - 1])

    def bfs(self, source: int, sink: int) -> bool:
        """Build level graph"""
        self.level = [-1] * self.n
        self.level[source] = 0
        queue = deque([source])

        while queue:
            node = queue.popleft()
            for next_node, cap, _ in self.graph[node]:
                if cap > 0 and self.level[next_node] < 0:
                    self.level[next_node] = self.level[node] + 1
                    queue.append(next_node)

        return self.level[sink] >= 0

    def dfs(self, node: int, sink: int, flow: int) -> int:
        """Find blocking flow"""
        if node == sink:
            return flow

        for i in range(self.iter[node], len(self.graph[node])):
            self.iter[node] = i
            next_node, cap, rev = self.graph[node][i]

            if cap > 0 and self.level[next_node] == self.level[node] + 1:
                d = self.dfs(next_node, sink, min(flow, cap))

                if d > 0:
                    self.graph[node][i][1] -= d
                    self.graph[next_node][rev][1] += d
                    return d

        return 0

    def max_flow(self, source: int, sink: int) -> int:
        """Compute maximum flow"""
        flow = 0

        while self.bfs(source, sink):
            self.iter = [0] * self.n

            while True:
                f = self.dfs(source, sink, float('inf'))
                if f == 0:
                    break
                flow += f

        return flow


# =============================================================================
# 4. Bipartite Matching
# =============================================================================

def bipartite_matching(n: int, m: int, edges: List[Tuple[int, int]]) -> int:
    """
    Maximum bipartite matching (simplified Hopcroft-Karp)
    n: number of left vertices, m: number of right vertices
    edges: [(left, right), ...]
    Time Complexity: O(E * sqrt(V))
    """
    # Build graph
    adj = defaultdict(list)
    for u, v in edges:
        adj[u].append(v)

    match_left = [-1] * n   # Left vertex matching
    match_right = [-1] * m  # Right vertex matching

    def dfs(u: int, visited: List[bool]) -> bool:
        for v in adj[u]:
            if visited[v]:
                continue
            visited[v] = True

            # Unmatched or can be rematched
            if match_right[v] == -1 or dfs(match_right[v], visited):
                match_left[u] = v
                match_right[v] = u
                return True

        return False

    matching = 0
    for u in range(n):
        visited = [False] * m
        if dfs(u, visited):
            matching += 1

    return matching


# =============================================================================
# 5. Minimum Cut
# =============================================================================

def min_cut(capacity: List[List[int]], source: int, sink: int) -> Tuple[int, List[int]]:
    """
    Minimum cut = Maximum flow
    Returns: (cut capacity, source-side vertex set)
    """
    n = len(capacity)
    residual = [row[:] for row in capacity]

    # Compute max flow using Edmonds-Karp
    def bfs_flow() -> bool:
        parent = [-1] * n
        visited = [False] * n
        visited[source] = True
        queue = deque([source])

        while queue:
            node = queue.popleft()
            for next_node in range(n):
                if not visited[next_node] and residual[node][next_node] > 0:
                    visited[next_node] = True
                    parent[next_node] = node
                    queue.append(next_node)

        if parent[sink] == -1:
            return False

        path_flow = float('inf')
        node = sink
        while node != source:
            prev = parent[node]
            path_flow = min(path_flow, residual[prev][node])
            node = prev

        node = sink
        while node != source:
            prev = parent[node]
            residual[prev][node] -= path_flow
            residual[node][prev] += path_flow
            node = prev

        return True

    while bfs_flow():
        pass

    # Find vertices reachable from source in residual graph
    visited = [False] * n
    queue = deque([source])
    visited[source] = True

    while queue:
        node = queue.popleft()
        for next_node in range(n):
            if not visited[next_node] and residual[node][next_node] > 0:
                visited[next_node] = True
                queue.append(next_node)

    # Compute cut capacity
    cut_capacity = 0
    source_side = []
    for i in range(n):
        if visited[i]:
            source_side.append(i)
            for j in range(n):
                if not visited[j] and capacity[i][j] > 0:
                    cut_capacity += capacity[i][j]

    return cut_capacity, source_side


# =============================================================================
# 6. Practical Problem: Task Assignment
# =============================================================================

def assign_tasks(workers: int, tasks: int, can_do: List[List[int]]) -> List[int]:
    """
    Assign tasks to workers (bipartite matching)
    can_do[i] = list of tasks worker i can perform
    Returns: assignment[worker] = assigned task (-1 if unassigned)
    """
    edges = []
    for worker, tasks_list in enumerate(can_do):
        for task in tasks_list:
            edges.append((worker, task))

    # Perform bipartite matching
    adj = defaultdict(list)
    for u, v in edges:
        adj[u].append(v)

    match_worker = [-1] * workers
    match_task = [-1] * tasks

    def dfs(u: int, visited: List[bool]) -> bool:
        for v in adj[u]:
            if visited[v]:
                continue
            visited[v] = True

            if match_task[v] == -1 or dfs(match_task[v], visited):
                match_worker[u] = v
                match_task[v] = u
                return True

        return False

    for u in range(workers):
        visited = [False] * tasks
        dfs(u, visited)

    return match_worker


# =============================================================================
# 7. Practical Problem: Maximum Edge-Disjoint Paths
# =============================================================================

def max_edge_disjoint_paths(n: int, edges: List[Tuple[int, int]], source: int, sink: int) -> int:
    """
    Maximum number of edge-disjoint paths
    Set each edge capacity = 1 and compute max flow
    """
    capacity = [[0] * n for _ in range(n)]

    for u, v in edges:
        capacity[u][v] = 1
        capacity[v][u] = 1  # For undirected graphs

    return edmonds_karp(capacity, source, sink)


# =============================================================================
# Tests
# =============================================================================

def main():
    print("=" * 60)
    print("Network Flow Examples")
    print("=" * 60)

    # Graph example:
    #     10      10
    #  0 ---> 1 ---> 3
    #  |      |      ^
    #  |10    |2     |10
    #  v      v      |
    #  2 ---> 4 ---> 5
    #     10      10

    # 1. Ford-Fulkerson
    print("\n[1] Ford-Fulkerson")
    capacity = [
        [0, 10, 10, 0, 0, 0],  # 0
        [0, 0, 2, 10, 0, 0],   # 1
        [0, 0, 0, 0, 10, 0],   # 2
        [0, 0, 0, 0, 0, 10],   # 3
        [0, 0, 0, 0, 0, 10],   # 4
        [0, 0, 0, 0, 0, 0]     # 5 (sink)
    ]
    flow = ford_fulkerson(capacity, 0, 5)
    print(f"    source=0, sink=5")
    print(f"    Maximum flow: {flow}")

    # 2. Edmonds-Karp
    print("\n[2] Edmonds-Karp")
    flow = edmonds_karp(capacity, 0, 5)
    print(f"    Maximum flow: {flow}")

    # 3. Dinic
    print("\n[3] Dinic's Algorithm")
    dinic = Dinic(6)
    dinic.add_edge(0, 1, 10)
    dinic.add_edge(0, 2, 10)
    dinic.add_edge(1, 2, 2)
    dinic.add_edge(1, 3, 10)
    dinic.add_edge(2, 4, 10)
    dinic.add_edge(3, 5, 10)
    dinic.add_edge(4, 5, 10)
    flow = dinic.max_flow(0, 5)
    print(f"    Maximum flow: {flow}")

    # 4. Bipartite Matching
    print("\n[4] Bipartite Matching")
    # Left: 0, 1, 2 (workers)
    # Right: 0, 1, 2 (tasks)
    edges = [(0, 0), (0, 1), (1, 0), (1, 2), (2, 1)]
    matching = bipartite_matching(3, 3, edges)
    print(f"    Edges: {edges}")
    print(f"    Maximum matching: {matching}")

    # 5. Minimum Cut
    print("\n[5] Minimum Cut")
    cut_cap, source_side = min_cut(capacity, 0, 5)
    print(f"    Minimum cut capacity: {cut_cap}")
    print(f"    Source-side vertices: {source_side}")

    # 6. Task Assignment
    print("\n[6] Task Assignment")
    can_do = [
        [0, 1],     # Worker 0: can do tasks 0, 1
        [1, 2],     # Worker 1: can do tasks 1, 2
        [0, 2]      # Worker 2: can do tasks 0, 2
    ]
    assignment = assign_tasks(3, 3, can_do)
    print(f"    Capabilities: {can_do}")
    print(f"    Assignment: {assignment}")

    # 7. Algorithm Comparison
    print("\n[7] Algorithm Complexity Comparison")
    print("    | Algorithm      | Time Complexity    | Notes              |")
    print("    |----------------|--------------------|--------------------|")
    print("    | Ford-Fulkerson | O(E * max_flow)    | DFS, integer cap   |")
    print("    | Edmonds-Karp   | O(V * E^2)         | BFS, stable        |")
    print("    | Dinic          | O(V^2 * E)         | Level graph        |")
    print("    | Bipartite      | O(E * sqrt(V))     | Dinic special case |")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()

File: examples/Algorithm/c/25_network_flow.c
/*
 * Network Flow
 * Ford-Fulkerson, Edmonds-Karp, Bipartite Matching, Min Cut
 *
 * Algorithms for finding maximum flow in a graph.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <stdbool.h>

#define MAX_V 1001
#define INF INT_MAX

/* =============================================================================
 * 1. Ford-Fulkerson (DFS-based)
 * ============================================================================= */

int capacity_ff[MAX_V][MAX_V];
bool visited_ff[MAX_V];

int dfs_ff(int u, int sink, int flow) {
    if (u == sink) return flow;
    visited_ff[u] = true;

    for (int v = 0; v < MAX_V; v++) {
        if (!visited_ff[v] && capacity_ff[u][v] > 0) {
            int min_flow = (flow < capacity_ff[u][v]) ? flow : capacity_ff[u][v];
            int result = dfs_ff(v, sink, min_flow);
            if (result > 0) {
                capacity_ff[u][v] -= result;
                capacity_ff[v][u] += result;
                return result;
            }
        }
    }
    return 0;
}

int ford_fulkerson(int source, int sink, int n) {
    int max_flow = 0;
    int flow;

    while (1) {
        memset(visited_ff, false, sizeof(visited_ff));
        flow = dfs_ff(source, sink, INF);
        if (flow == 0) break;
        max_flow += flow;
    }

    return max_flow;
}

/* =============================================================================
 * 2. Edmonds-Karp (BFS-based, O(VE^2))
 * ============================================================================= */

typedef struct {
    int** capacity;
    int** flow;
    int n;
} FlowNetwork;

FlowNetwork* fn_create(int n) {
    FlowNetwork* fn = malloc(sizeof(FlowNetwork));
    fn->n = n;
    fn->capacity = malloc(n * sizeof(int*));
    fn->flow = malloc(n * sizeof(int*));
    for (int i = 0; i < n; i++) {
        fn->capacity[i] = calloc(n, sizeof(int));
        fn->flow[i] = calloc(n, sizeof(int));
    }
    return fn;
}

void fn_free(FlowNetwork* fn) {
    for (int i = 0; i < fn->n; i++) {
        free(fn->capacity[i]);
        free(fn->flow[i]);
    }
    free(fn->capacity);
    free(fn->flow);
    free(fn);
}

void fn_add_edge(FlowNetwork* fn, int u, int v, int cap) {
    fn->capacity[u][v] += cap;
}

int edmonds_karp(FlowNetwork* fn, int source, int sink) {
    int max_flow = 0;
    int* parent = malloc(fn->n * sizeof(int));
    int* queue = malloc(fn->n * sizeof(int));

    while (1) {
        memset(parent, -1, fn->n * sizeof(int));
        parent[source] = source;

        /* BFS */
        int front = 0, rear = 0;
        queue[rear++] = source;

        while (front < rear && parent[sink] == -1) {
            int u = queue[front++];
            for (int v = 0; v < fn->n; v++) {
                if (parent[v] == -1 &&
                    fn->capacity[u][v] - fn->flow[u][v] > 0) {
                    parent[v] = u;
                    queue[rear++] = v;
                }
            }
        }

        if (parent[sink] == -1) break;

        /* Find minimum capacity along the path */
        int path_flow = INF;
        for (int v = sink; v != source; v = parent[v]) {
            int u = parent[v];
            int residual = fn->capacity[u][v] - fn->flow[u][v];
            if (residual < path_flow) path_flow = residual;
        }

        /* Update flow */
        for (int v = sink; v != source; v = parent[v]) {
            int u = parent[v];
            fn->flow[u][v] += path_flow;
            fn->flow[v][u] -= path_flow;
        }

        max_flow += path_flow;
    }

    free(parent);
    free(queue);
    return max_flow;
}

/* =============================================================================
 * 3. Bipartite Matching
 * ============================================================================= */

typedef struct {
    int** adj;
    int* adj_size;
    int* match_left;
    int* match_right;
    bool* visited;
    int left_n;
    int right_n;
} BipartiteGraph;

BipartiteGraph* bg_create(int left_n, int right_n) {
    BipartiteGraph* bg = malloc(sizeof(BipartiteGraph));
    bg->left_n = left_n;
    bg->right_n = right_n;
    bg->adj = malloc(left_n * sizeof(int*));
    bg->adj_size = calloc(left_n, sizeof(int));
    for (int i = 0; i < left_n; i++) {
        bg->adj[i] = malloc(right_n * sizeof(int));
    }
    bg->match_left = malloc(left_n * sizeof(int));
    bg->match_right = malloc(right_n * sizeof(int));
    bg->visited = malloc(right_n * sizeof(bool));
    memset(bg->match_left, -1, left_n * sizeof(int));
    memset(bg->match_right, -1, right_n * sizeof(int));
    return bg;
}

void bg_free(BipartiteGraph* bg) {
    for (int i = 0; i < bg->left_n; i++) {
        free(bg->adj[i]);
    }
    free(bg->adj);
    free(bg->adj_size);
    free(bg->match_left);
    free(bg->match_right);
    free(bg->visited);
    free(bg);
}

void bg_add_edge(BipartiteGraph* bg, int left, int right) {
    bg->adj[left][bg->adj_size[left]++] = right;
}

bool bg_dfs(BipartiteGraph* bg, int u) {
    for (int i = 0; i < bg->adj_size[u]; i++) {
        int v = bg->adj[u][i];
        if (bg->visited[v]) continue;
        bg->visited[v] = true;

        if (bg->match_right[v] == -1 || bg_dfs(bg, bg->match_right[v])) {
            bg->match_left[u] = v;
            bg->match_right[v] = u;
            return true;
        }
    }
    return false;
}

int bipartite_matching(BipartiteGraph* bg) {
    int matching = 0;

    for (int u = 0; u < bg->left_n; u++) {
        memset(bg->visited, false, bg->right_n * sizeof(bool));
        if (bg_dfs(bg, u)) {
            matching++;
        }
    }

    return matching;
}

/* =============================================================================
 * 4. Hopcroft-Karp (O(E*sqrt(V)))
 * ============================================================================= */

int* dist_hk;
int* match_left_hk;
int* match_right_hk;
int** adj_hk;
int* adj_size_hk;
int left_n_hk, right_n_hk;

bool hk_bfs(void) {
    int* queue = malloc((left_n_hk + 1) * sizeof(int));
    int front = 0, rear = 0;

    for (int u = 0; u < left_n_hk; u++) {
        if (match_left_hk[u] == -1) {
            dist_hk[u] = 0;
            queue[rear++] = u;
        } else {
            dist_hk[u] = INF;
        }
    }

    bool found = false;
    while (front < rear) {
        int u = queue[front++];
        for (int i = 0; i < adj_size_hk[u]; i++) {
            int v = adj_hk[u][i];
            int next = match_right_hk[v];
            if (next == -1) {
                found = true;
            } else if (dist_hk[next] == INF) {
                dist_hk[next] = dist_hk[u] + 1;
                queue[rear++] = next;
            }
        }
    }

    free(queue);
    return found;
}

bool hk_dfs(int u) {
    for (int i = 0; i < adj_size_hk[u]; i++) {
        int v = adj_hk[u][i];
        int next = match_right_hk[v];
        if (next == -1 || (dist_hk[next] == dist_hk[u] + 1 && hk_dfs(next))) {
            match_left_hk[u] = v;
            match_right_hk[v] = u;
            return true;
        }
    }
    dist_hk[u] = INF;
    return false;
}

int hopcroft_karp(int left_n, int right_n, int** adj, int* adj_size) {
    left_n_hk = left_n;
    right_n_hk = right_n;
    adj_hk = adj;
    adj_size_hk = adj_size;

    dist_hk = malloc(left_n * sizeof(int));
    match_left_hk = malloc(left_n * sizeof(int));
    match_right_hk = malloc(right_n * sizeof(int));

    memset(match_left_hk, -1, left_n * sizeof(int));
    memset(match_right_hk, -1, right_n * sizeof(int));

    int matching = 0;
    while (hk_bfs()) {
        for (int u = 0; u < left_n; u++) {
            if (match_left_hk[u] == -1 && hk_dfs(u)) {
                matching++;
            }
        }
    }

    free(dist_hk);
    free(match_left_hk);
    free(match_right_hk);
    return matching;
}

/* =============================================================================
 * 5. Min Cut
 * ============================================================================= */

void find_min_cut(FlowNetwork* fn, int source, int* reachable, int* cut_size) {
    bool* visited = calloc(fn->n, sizeof(bool));
    int* queue = malloc(fn->n * sizeof(int));
    int front = 0, rear = 0;

    visited[source] = true;
    queue[rear++] = source;

    *cut_size = 0;

    while (front < rear) {
        int u = queue[front++];
        reachable[(*cut_size)++] = u;
        for (int v = 0; v < fn->n; v++) {
            if (!visited[v] && fn->capacity[u][v] - fn->flow[u][v] > 0) {
                visited[v] = true;
                queue[rear++] = v;
            }
        }
    }

    free(visited);
    free(queue);
}

/* =============================================================================
 * Test
 * ============================================================================= */

int main(void) {
    printf("============================================================\n");
    printf("Network Flow Examples\n");
    printf("============================================================\n");

    /* 1. Edmonds-Karp */
    printf("\n[1] Edmonds-Karp Algorithm\n");
    FlowNetwork* fn = fn_create(6);
    fn_add_edge(fn, 0, 1, 16);
    fn_add_edge(fn, 0, 2, 13);
    fn_add_edge(fn, 1, 2, 10);
    fn_add_edge(fn, 1, 3, 12);
    fn_add_edge(fn, 2, 1, 4);
    fn_add_edge(fn, 2, 4, 14);
    fn_add_edge(fn, 3, 2, 9);
    fn_add_edge(fn, 3, 5, 20);
    fn_add_edge(fn, 4, 3, 7);
    fn_add_edge(fn, 4, 5, 4);

    printf("    Graph:\n");
    printf("      0 -> 1 (16), 0 -> 2 (13)\n");
    printf("      1 -> 2 (10), 1 -> 3 (12)\n");
    printf("      2 -> 1 (4),  2 -> 4 (14)\n");
    printf("      3 -> 2 (9),  3 -> 5 (20)\n");
    printf("      4 -> 3 (7),  4 -> 5 (4)\n");
    printf("    Max flow (0->5): %d\n", edmonds_karp(fn, 0, 5));
    fn_free(fn);

    /* 2. Bipartite Matching */
    printf("\n[2] Bipartite Matching\n");
    BipartiteGraph* bg = bg_create(4, 4);
    /* Worker(0-3) to Task(0-3) matching */
    bg_add_edge(bg, 0, 0);
    bg_add_edge(bg, 0, 1);
    bg_add_edge(bg, 1, 0);
    bg_add_edge(bg, 1, 2);
    bg_add_edge(bg, 2, 1);
    bg_add_edge(bg, 2, 2);
    bg_add_edge(bg, 3, 2);
    bg_add_edge(bg, 3, 3);

    printf("    Worker-Task connections:\n");
    printf("      Worker0: Task0, Task1\n");
    printf("      Worker1: Task0, Task2\n");
    printf("      Worker2: Task1, Task2\n");
    printf("      Worker3: Task2, Task3\n");
    printf("    Maximum matching: %d\n", bipartite_matching(bg));

    printf("    Matching result:\n");
    for (int i = 0; i < bg->left_n; i++) {
        if (bg->match_left[i] != -1) {
            printf("      Worker%d -> Task%d\n", i, bg->match_left[i]);
        }
    }
    bg_free(bg);

    /* 3. Minimum Vertex Cover */
    printf("\n[3] Minimum Vertex Cover (Bipartite Graph)\n");
    printf("    Konig's theorem: Max matching = Min vertex cover\n");
    printf("    Min vertex cover of above example: 4\n");

    /* 4. Ford-Fulkerson */
    printf("\n[4] Ford-Fulkerson (Simple Example)\n");
    memset(capacity_ff, 0, sizeof(capacity_ff));
    capacity_ff[0][1] = 10;
    capacity_ff[0][2] = 10;
    capacity_ff[1][2] = 2;
    capacity_ff[1][3] = 4;
    capacity_ff[1][4] = 8;
    capacity_ff[2][4] = 9;
    capacity_ff[3][5] = 10;
    capacity_ff[4][3] = 6;
    capacity_ff[4][5] = 10;

    printf("    Max flow (0->5): %d\n", ford_fulkerson(0, 5, 6));

    /* 5. Applications */
    printf("\n[5] Network Flow Applications\n");
    printf("    - Bipartite matching: task assignment, marriage problem\n");
    printf("    - Min cut: network partitioning\n");
    printf("    - Maximum independent set (bipartite graph)\n");
    printf("    - Project selection problem\n");
    printf("    - Circulation flow\n");

    /* 6. Complexity */
    printf("\n[6] Complexity\n");
    printf("    | Algorithm        | Time          |\n");
    printf("    |------------------|---------------|\n");
    printf("    | Ford-Fulkerson   | O(E x f)      |\n");
    printf("    | Edmonds-Karp     | O(VE^2)       |\n");
    printf("    | Dinic            | O(V^2E)       |\n");
    printf("    | Bipartite (DFS)  | O(VE)         |\n");
    printf("    | Hopcroft-Karp    | O(E*sqrt(V))  |\n");
    printf("    f: maximum flow\n");

    printf("\n============================================================\n");

    return 0;
}
